services/continuous_querier/service.go

```diff
@@ -1 +1,368 @@
 package continuous_querier
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/influxdb/influxdb/cluster"
+	"github.com/influxdb/influxdb/influxql"
+	"github.com/influxdb/influxdb/meta"
+	"github.com/influxdb/influxdb/tsdb"
+)
+
+const (
+	// When planning a select statement, passing zero tells it not to chunk results. Only applies to raw queries
+	NoChunkingSize = 0
+)
+
+// queryExecutor is an internal interface to make testing easier.
+type queryExecutor interface {
+	ExecuteQuery(query *influxql.Query, database string, chunkSize int) (<-chan *influxql.Result, error)
+}
+
+// metaStore is an internal interface to make testing easier.
+type metaStore interface {
+	IsLeader() bool
+	Databases() ([]meta.DatabaseInfo, error)
+}
+
+// pointsWriter is an internal interface to make testing easier.
+type pointsWriter interface {
+	WritePoints(p *cluster.WritePointsRequest) error
+}
+
+// Service manages continuous query execution.
+type Service struct {
+	MetaStore     metaStore
+	QueryExecutor queryExecutor
+	PointsWriter  pointsWriter
+	Config        *Config
+	RunInterval   time.Duration
+	Logger        *log.Logger
+	// lastRuns maps CQ name to last time it was run.
+	lastRuns map[string]time.Time
+	stop     chan struct{}
+	wg       *sync.WaitGroup
+}
+
+// NewService returns a new instance of Service.
+func NewService(c Config) *Service {
+	s := &Service{
+		Config:      &c,
+		RunInterval: time.Second,
+		Logger:      log.New(os.Stderr, "[continuous_querier] ", log.LstdFlags),
+		lastRuns:    map[string]time.Time{},
+	}
+	return s
+}
+
+// Open starts the service.
+func (s *Service) Open() error {
+	if s.stop != nil {
+		return nil
+	}
+
+	assert(s.MetaStore != nil, "MetaStore is nil")
+	assert(s.QueryExecutor != nil, "QueryExecutor is nil")
+	assert(s.PointsWriter != nil, "PointsWriter is nil")
+
+	s.stop = make(chan struct{})
+	s.wg = &sync.WaitGroup{}
+	s.wg.Add(1)
+	go s.backgroundLoop()
+	return nil
+}
+
+// Close stops the service.
+func (s *Service) Close() error {
+	if s.stop == nil {
+		return nil
+	}
+	close(s.stop)
+	s.wg.Wait()
+	s.wg = nil
+	s.stop = nil
+	return nil
+}
+
+// backgroundLoop runs on a go routine and periodically executes CQs.
+func (s *Service) backgroundLoop() {
+	defer s.wg.Done()
+	for {
+		select {
+		case <-s.stop:
+			return
+		case <-time.After(s.RunInterval):
+			if s.MetaStore.IsLeader() {
+				s.runContinuousQueries()
+			}
+		}
+	}
+}
+
+// runContinuousQueries gets CQs from the meta store and runs them.
+func (s *Service) runContinuousQueries() {
+	// Get list of all databases.
+	dbs, err := s.MetaStore.Databases()
+	if err != nil {
+		s.Logger.Println("error getting databases")
+		return
+	}
+	// Loop through all databases executing CQs.
+	for _, db := range dbs {
+		// TODO: distribute across nodes
+		for _, cq := range db.ContinuousQueries {
+			if err := s.ExecuteContinuousQuery(&db, &cq); err != nil {
+				s.Logger.Println(err)
```

Comment:
Should this be `Printf` so you can enhance the error message? E.g. "CQ run failure due to error:....".